File: scripts/extract-keybindings.py
#!/usr/bin/env python3
"""Extract keybindings from Neovim Lua configuration files."""

import os
import re
import glob
from collections import defaultdict

def extract_keybindings_from_file(filepath):
    """Extract keybindings from a single Lua file."""
    keybindings = []

    try:
        with open(filepath, 'r', encoding='utf-8') as f:
            content = f.read()

        # Pattern to match vim.keymap.set calls
        pattern = r'vim\.keymap\.set\s*\(\s*["\']([^"\']*)["\'],\s*["\']([^"\']*)["\'][^}]*desc\s*=\s*["\']([^"\']*)["\']'

        matches = re.findall(pattern, content, re.MULTILINE)

        for mode, key, desc in matches:
            keybindings.append({
                'file': os.path.basename(filepath),
                'mode': mode,
                'key': key,
                'description': desc
            })

    except Exception as e:
        print(f"Error processing {filepath}: {e}")

    return keybindings

def main():
    """Generate keybinding documentation."""

    # Find all Lua files
    lua_files = glob.glob('lua/**/*.lua', recursive=True)
    lua_files.extend(glob.glob('*.lua'))

    all_keybindings = []

    for filepath in lua_files:
        keybindings = extract_keybindings_from_file(filepath)
        all_keybindings.extend(keybindings)

    # Group by mode
    keybindings_by_mode = defaultdict(list)
    for kb in all_keybindings:
        keybindings_by_mode[kb['mode']].append(kb)

    # Generate markdown
    os.makedirs('docs', exist_ok=True)

    with open('docs/KEYBINDINGS.md', 'w') as f:
        f.write("# Keybindings Reference\n\n")
        f.write("*Auto-generated from configuration files*\n\n")

        # Mode descriptions
        mode_descriptions = {
            'n': 'Normal Mode',
            'i': 'Insert Mode',
            'v': 'Visual Mode',
            'x': 'Visual Block Mode',
            't': 'Terminal Mode',
            'c': 'Command Mode'
        }

        for mode in sorted(keybindings_by_mode.keys()):
            mode_desc = mode_descriptions.get(mode, f"Mode '{mode}'")
            f.write(f"## {mode_desc}\n\n")
            f.write("| Key | Description | Source |\n")
            f.write("|-----|-------------|--------|\n")

            # Sort by key
            sorted_bindings = sorted(keybindings_by_mode[mode], key=lambda x: x['key'])

            for kb in sorted_bindings:
                key = kb['key'].replace('|', '\\|')  # Escape pipes for markdown
                desc = kb['description'].replace('|', '\\|')
                source = kb['file'].replace('.lua', '')
                f.write(f"| `{key}` | {desc} | {source} |\n")

            f.write("\n")

    print(f"Generated keybinding documentation with {len(all_keybindings)} keybindings")

if __name__ == '__main__':
    main()

File: scripts/extract-keybindings.lua
-- extract-keybindings.lua - generate docs/KEYBINDINGS.md from the config.
--
-- Run with: nvim -l scripts/extract-keybindings.lua
--
-- Parses every Lua file with tree-sitter, collects `*.keymap.set` calls that
-- have a string key and a `desc`, and writes a markdown reference grouped by
-- mode with clickable source links.

local script_dir = arg[0]:match("(.*[/\\])") or "./"
package.path = script_dir .. "lib/?.lua;" .. package.path
local docgen = require("docgen")

local function escape(s)
  return (s:gsub("|", "\\|"))
end

local mode_names = {
  n = "Normal Mode",
  i = "Insert Mode",
  v = "Visual Mode",
  x = "Visual Block Mode",
  s = "Select Mode",
  o = "Operator-pending Mode",
  t = "Terminal Mode",
  c = "Command Mode",
  [""] = "Normal, Visual, Operator-pending",
}

local files = {}
vim.list_extend(files, docgen.lua_files("lua", "**/*.lua"))
vim.list_extend(files, docgen.lua_files(".", "*.lua"))
table.sort(files)

-- Collect keymaps grouped by mode.
local by_mode = {}
local total = 0
for _, file in ipairs(files) do
  local src = docgen.read(file)
  if src then
    local rel = file:gsub("^%./", "")
    for _, km in ipairs(docgen.extract_keymaps(src)) do
      by_mode[km.mode] = by_mode[km.mode] or {}
      table.insert(by_mode[km.mode], { key = km.key, desc = km.desc, src = rel, line = km.line })
      total = total + 1
    end
  end
end

local modes = vim.tbl_keys(by_mode)
table.sort(modes)

local out = {}
table.insert(out, "# Keybindings Reference")
table.insert(out, "")
table.insert(out, "_Auto-generated from configuration files. Do not edit manually._")
table.insert(out, "")

for _, mode in ipairs(modes) do
  table.insert(out, "## " .. (mode_names[mode] or ("Mode '" .. mode .. "'")))
  table.insert(out, "")
  table.insert(out, "| Key | Description | Source |")
  table.insert(out, "|-----|-------------|--------|")
  local rows = by_mode[mode]
  table.sort(rows, function(a, b)
    if a.key == b.key then
      return a.src < b.src
    end
    return a.key < b.key
  end)
  for _, r in ipairs(rows) do
    local source = string.format("[%s#L%d](%s#L%d)", r.src, r.line, r.src, r.line)
    table.insert(out, string.format("| `%s` | %s | %s |", escape(r.key), escape(r.desc), source))
  end
  table.insert(out, "")
end

docgen.write("docs/KEYBINDINGS.md", table.concat(out, "\n"))
print(string.format("Generated docs/KEYBINDINGS.md with %d keybindings", total))
